### PR TITLE
Unetrpp scaled_dot_product_attention + various fixes

### PR DESCRIPTION
**Pull request by colon3ltocard**
* Adds proj_size as a setting to unetr++, all layers now use 64 size instead of [64, 64, 64, 32]
* Adds two booleans to enable/disable [torch's scaled_dot_product_attention](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html)  for unetrpp.
* Fixes observer and bin/test.py to work the same with both singledevice and multidevice testing/training
* Makes _HalfUnet usable as a submodel/component in other architectures

Note: in order to use scaled_dot_product_attention for spatial attention in the EPA block I had to slightly modify the way the matrices are computed and used: q_shared is now projected, the result is back projected with a linear transformation to the right size.

For observer modification I tested with the following commands :

```bash
runai gpu_play 1
 runai exec_gpu python bin/train.py --dataset dummy --model unetrpp --epochs 1
```

```bash
runai gpu_play 2
 runai exec_gpu python bin/train.py --dataset dummy --model unetrpp --epochs 1
```

Results are very close between the two implementations:

image

image

Test scores with scaled_dot

Test_rmse_161024_linear_up_scaled_dot

Test scores without

Test_rmse_161024_linear_up

### CODE REVIEW

File: py4cast/models/vision/unetrpp.py [hunk 1]
```diff
@@ -125,17 +126,18 @@ def __init__(
...
+        use_scaled_dot_product_SA: bool = True,
     ) -> None:
         """
         Args:
             input_size: the size of the input for each stage.
             hidden_size: dimension of hidden layer.
-            proj_size: projection size for keys and values in the spatial attention module.
             num_heads: number of attention heads.
             dropout_rate: faction of the input units to drop.
             pos_embed: bool argument to determine if positional embedding is used.
```
Comment (flyIchtus) [inline py4cast/models/vision/unetrpp.py:143]: Can you please update doc with the added bools ?

File: py4cast/observer.py [hunk 2]
```diff
@@ -422,7 +376,7 @@ def update(
                 spatial_loss, "b t (x y) -> b t x y ", x=obj.grid_shape[0]
             )
         self.spatial_loss_maps.append(
-            reduce(obj, spatial_loss).unsqueeze(0)
+            obj.trainer.strategy.reduce(spatial_loss, reduce_op="mean").cpu()
```
Comment (flyIchtus) [inline py4cast/observer.py:379]: What about the `unsqueeze` op  that has been erased ? is it voluntary ?

File: py4cast/models/vision/unetrpp.py (hunk 1 above)
Comment (colon3ltocard) [reply py4cast/models/vision/unetrpp.py:143]: Done

File: py4cast/models/vision/unetrpp.py [hunk 3]
```diff
@@ -266,21 +291,38 @@ def forward(self, x):
...
+        else:
+            attn_CA = (q_shared @ k_shared.transpose(-2, -1)) * self.temperature
+            attn_CA = attn_CA.softmax(dim=-1)
+            attn_CA = self.attn_drop(attn_CA)
+            x_CA = attn_CA @ v_CA
+
+        x_CA = x_CA.permute(0, 3, 1, 2).reshape(B, N, C)
+
+        if self.use_scaled_dot_product_SA:
+            q_shared_projected = torch.einsum("bhdn,nk->bhdk", q_shared, self.EF)
```
Comment (flyIchtus) [inline py4cast/models/vision/unetrpp.py:312]: The `self.temperature` play no role with sdpa but they are used in the other case : is this a desired behaviour ?
(like, adding self.temperature as the "scale" parameter ?)

File: py4cast/models/vision/unetrpp.py [hunk 4]
```diff
@@ -266,21 +291,38 @@ def forward(self, x):
...
+                q_shared_projected.permute(0, 1, 3, 2),
+                k_shared_projected.permute(0, 1, 3, 2),
+                v_SA_projected.permute(0, 1, 3, 2),
+                dropout_p=self.attn_drop_2.p,
+            )
+            x_SA = torch.einsum("bhdn,nk->bhdk", x_SA.permute(0, 1, 3, 2), self.FE)
+
+        else:
+            attn_SA = (
+                q_shared.permute(0, 1, 3, 2) @ k_shared_projected
```
Comment (flyIchtus) [inline py4cast/models/vision/unetrpp.py:323]: why is `q_shared_projected` computed only in sdpa ?

**Review by flyIchtus — COMMENTED** [review]
Nice job !!
Just a few pending questions : maybe it's just me not familiar enough with the code. If you deem them irrelevant, just discard them and I will happily resolve.

File: py4cast/models/vision/conv.py
Comment (cbovalo) [inline py4cast/models/vision/conv.py:1]: What is the idea of having distinct classes _HalfUnet and HalfUnet ?

File: py4cast/models/vision/unetrpp.py [hunk 5]
```diff
@@ -243,16 +253,31 @@ def __init__(
...
+
+        # Matrix index, Batch, Head, Dimensions, Features
         qkvv = qkvv.permute(2, 0, 3, 1, 4)
+
+        # Batch, Head, Dimensions, Features
         q_shared, k_shared, v_CA, v_SA = qkvv[0], qkvv[1], qkvv[2], qkvv[3]
 
+        # Batch, Head, Features, Dimensions
         q_shared = q_shared.transpose(-2, -1)
         k_shared = k_shared.transpose(-2, -1)
```
[the remark this answers is not in the document]
Comment (cbovalo) [reply py4cast/models/vision/unetrpp.py:287]: Could we remove the transpose() calls and add them when doing the permute(): `qkvv = qkvv.permute(2, 0, 3, 4, 1)` instead the original one ?

File: py4cast/observer.py (hunk 2 above)
Comment (colon3ltocard) [reply py4cast/observer.py:379]: Yes it is voluntary.

Well there was a flatten(0, 1) in one condition (Distributed) of the former reduce function followed by this unsqueeze(0). Made no sense. 

In SingleDevice there was no flatten(0, 1) and there was an exception raised in the plot code later on because there was an extra dimension in the tensors.

File: py4cast/models/vision/unetrpp.py [hunk 6]
```diff
@@ -297,14 +344,16 @@ def __init__(
         self,
         input_size=[32 * 32 * 32, 16 * 16 * 16, 8 * 8 * 8, 4 * 4 * 4],
```
Comment (cbovalo) [inline py4cast/models/vision/unetrpp.py:345]: This is out of the scope of this PR but could we imagine using powers of 2 to define `input_size` and `dims` ?

File: py4cast/models/vision/unetrpp.py (hunk 3 above)
Comment (colon3ltocard) [reply py4cast/models/vision/unetrpp.py:312]: I tried passing the temperature as the scale argument of sdpa but it doesn"t accept trainable parameters. This is a design inherited from the original paper implementation so yes it is desired in the sense that it is the best compromise i found to use sdpa.

File: py4cast/models/vision/conv.py
Comment (colon3ltocard) [reply py4cast/models/vision/conv.py:1]: I can reuse _HalfUnet as a submodel/component whereas HalfUnet has the "py4cast" tricks/adaptations in its forward regarding feature ordering. I did that because HU has a small receptive field and I wanted to use it as submodel, for local and fine-grained computations, in a broader architecture.

File: py4cast/models/vision/unetrpp.py (hunk 6 above)
Comment (colon3ltocard) [reply py4cast/models/vision/unetrpp.py:345]: We could, it is coming straight from the original repo/implementation.

File: py4cast/models/vision/unetrpp.py [hunk 7]
```diff
@@ -243,16 +253,31 @@ def __init__(
...
 
+        if self.use_scaled_dot_product_SA:
+            # we need an extra set of weights for the scaled dot product
+            self.FE = nn.Parameter(init_(torch.zeros(proj_size, input_size)))
+        else:
+            # the original unetr++ uses two parameters for scaling the dot products
+            self.temperature2 = nn.Parameter(torch.ones(num_heads, 1, 1))
+
+        if not self.use_scaled_dot_product_CA:
+            self.temperature = nn.Parameter(torch.ones(num_heads, 1, 1))
```
[the remark this answers is not in the document]
Comment (colon3ltocard) [reply py4cast/models/vision/unetrpp.py:269]: Nope. It is either instanciated and used or not instanciated and not used. Torch raises an exception if some parameters weights are left unused during the training process.

File: py4cast/models/vision/unetrpp.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (colon3ltocard) [reply py4cast/models/vision/unetrpp.py:287]: We could. For this PR I choose to stick as close as possible to the original implementation while adding the sdpa to minimize risks and facilitate comparison. I think we probably should have a got at simplifying/optmising all these in a dedicated PR (also because the current PR comes with 5 days of training to validate the results)

File: py4cast/models/vision/unetrpp.py (hunk 4 above)
Comment (colon3ltocard) [reply py4cast/models/vision/unetrpp.py:323]: passing the matrices (q_shared, k_shared_projected, ...) used in the original implementation for spatial attention directly to sdpa leads to exceptions because the shape weren't right. So I decided to also project q_shared in the same subspace and permute to have a "spatial attention", "after projection" done by sdpa. If you have other ideas I am happy to try.

**Review by cbovalo — APPROVED** [review]

File: py4cast/models/vision/unetrpp.py (hunk 3 above)
Comment (flyIchtus) [reply py4cast/models/vision/unetrpp.py:312]: What about multipliying `q_shared` / `k_shared` by `torch.rsqrt(self.temperature)` beforehand and keep the scale kwargs to 1.0 ?

Comment (colon3ltocard) [reply py4cast/models/vision/unetrpp.py:312]: We could try yes, also compare the results with and without it, and if it doesn't improve the performances, we could probably get rid of it altogether ?

**Review by A669015 — APPROVED** [review]

File: py4cast/models/vision/unetrpp.py (hunk 3 above)
Comment (flyIchtus) [reply py4cast/models/vision/unetrpp.py:312]: ok, later pr

File: py4cast/models/vision/unetrpp.py (hunk 4 above)
Comment (flyIchtus) [reply py4cast/models/vision/unetrpp.py:323]: Ok, so after discussion with PR owner, here's the pain:

-  sdpa requires same size q /k matrices, so, does not like k_projected and q_shared
- the current solution actually increases costs due to `q_shared_projected@k_shared_projected.transpose` which is O(D²) (with D the spatial dim). Any monkey-patch will result in the O(D²) popping somewhere.
- there might be a final shape mismatch in `x_SA` (not happening because of the channels/dims consistent evolution)

Comment (colon3ltocard) [reply py4cast/models/vision/unetrpp.py]: Done. I launched a cheap experiment to compare results. thanks @flyIchtus

**Review by flyIchtus — APPROVED** [review]
